lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+-- @usage colorful.color("#123") + colorful.color("#321") -- returns #444444
+color_mt.__add = function(lhs, rhs)
+    if type(rhs) ~= "table" then return nil end
+    return color.from.rgba(
+        (lhs.red + rhs.red) * 255,
+        (lhs.green + rhs.green) * 255,
+        (lhs.blue + rhs.blue) * 255,
+        (lhs.alpha + rhs.alpha) * 255
+    )
+end
+
+--- Subtract the RGBA values of two colors
+-- @function color.__sub
+-- @tparam lhs beautiful.colorful.color A color object
+-- @tparam rhs beautiful.colorful.color A color object
+-- @treturn beautiful.colorful.color A new color object with RGBA values
+-- subtracted
+-- @usage colorful.color("#123") - colorful.color("#321") -- returns #000022
+color_mt.__sub = function(lhs, rhs)
+    if type(rhs) ~= "table" then return nil end
+    return color.from.rgba(
+        (lhs.red - rhs.red) * 255,
+        (lhs.green - rhs.green) * 255,
+        (lhs.blue - rhs.blue) * 255,
+        (lhs.alpha - rhs.alpha) * 255
+    )
+end
+
+--- Compare the lightness value of two colors
+-- @function color.__lt
+-- @tparam lhs beautiful.colorful.color A color object
+-- @tparam rhs beautiful.colorful.color A color object
+-- @treturn boolean True if lhs lightness is less than rhs lightness
+-- @usage colorful.color("#ffcc99") < colorful.color("#aabbcc") -- returns false (0.8 < 0.73333)
+color_mt.__lt = function(lhs, rhs)
+    if type(rhs) ~= "table" then return nil end
+    return lhs.lightness < rhs.lightness
+end
+
+--- Compare the lightness value of two colors
+-- @function color.__gt
+-- @tparam lhs beautiful.colorful.color A color object
+-- @tparam rhs beautiful.colorful.color A color object
+-- @treturn boolean True if lhs lightness is greater than rhs lightness
+-- @usage colorful.color("#ffcc99") > colorful.color("#aabbcc") -- returns true (0.8 > 0.73333)
+color_mt.__gt = function(lhs, rhs)
+    if type(rhs) ~= "table" then return nil end
+    return lhs.lightness > rhs.lightness
+end
+
+--- A color object
+-- @type color
+colorful.color = color
+
+setmetatable(color, {
+    __call = function(_,...) return color.new(...) end,
+})
+-- {{{ Color changing wrapper functions
+
+--- Wrappers for Color object functions
+-- @section wrappers
+
+--- Mix two colors with optional mix amount
+-- @param color1 A CSS color compatible string, decimal color value, or a table
+-- containing 3 or 4 values.
+-- @param color2 A CSS color compatible string, decimal color value, or a table
+-- containing 3 or 4 values.
+-- @tparam number amount A number [0,1] of how much of color1 should be mixed
+-- with color2.
+-- @treturn string The string representation of the mixed color in `#rrggbbaa`
+-- format.
+-- @see beautiful.colorful.color.mix
+-- @usage local mix_str = colorful.mix("#ffcc99", "#99ccff", 0.5)
+function colorful.mix(color1, color2, amount)
+    local c1 = color.new(color1)
+    local c2 = color.new(color2)
+    if c1 == nil or c2 == nil then return nil end
```

Comment:
Would it make sense to have this cause errors instead of returning `nil`? `nil`s might mean that some background color is just silently ignored, which is a bit more confusing than "I get an error with a traceback".

Reply:
To me I'd like it to just return nil as it shows something went wrong and the user can catch that and use a fallback or error on their own if it's important enough. But if you think error'ing is better here then that's fine.